common/monitor.go

```diff
@@ -1,158 +1,156 @@
-	alerts     []chan FDStatus // Channels to send minutes to
-	polls      int64
-	info       *factom.CurrentMinuteInfo
-	status     string
+var monitor *Monitor
+var once sync.Once
+
+// GetMonitor returns the singleton instance of the Factomd Monitor
+func GetMonitor() *Monitor {
+	once.Do(func() {
+		monitor = new(Monitor)
+		monitor.errors = []chan error{}
+		monitor.listeners = []chan MonitorEvent{}
+		go monitor.poll()
+	})
+	return monitor
 }
 
-func (f *FactomdMonitor) GetAlert() chan FDStatus {
-	f.mutex.Lock()
-	defer f.mutex.Unlock()
-	alert := make(chan FDStatus, 10)
-	f.alerts = append(f.alerts, alert)
-	return alert
+// Monitor polls a factomd node and sends alerts whenever the block height or minute changes
+type Monitor struct {
+	polls   int64
+	timeout time.Duration
+
+	listenerMutex sync.Mutex
+	listeners     []chan MonitorEvent // Channels to send minutes to
+	errorMutex    sync.Mutex
+	errors        []chan error
+}
+
+// MonitorEvent is the data sent to all listeners when being notified
+type MonitorEvent struct {
+	Minute int64
+	Dbht   int32
+}
+
+// Listener spawns a new listening channel that will receive updates of height or minute changes
+func (f *Monitor) Listener() <-chan MonitorEvent {
+	f.listenerMutex.Lock()
+	defer f.listenerMutex.Unlock()
+
+	listener := make(chan MonitorEvent, 10)
+	f.listeners = append(f.listeners, listener)
+	return listener
+}
+
+// ErrorListener spawns a new listening channel that will receive errors that have occurred
+func (f *Monitor) ErrorListener() <-chan error {
+	f.errorMutex.Lock()
+	defer f.errorMutex.Unlock()
+
+	listener := make(chan error, 1)
+	f.errors = append(f.errors, listener)
+	return listener
 }
 
-// GetBlockTime
-// Returns the blocktime in seconds.  All blocks are divided into 10 "minute" sections.  But if the blocktime
-// is not 600 seconds, then a minute = the blocktime/10
-func (f *FactomdMonitor) GetBlockTime() int64 {
-	f.mutex.Lock()
-	defer f.mutex.Unlock()
-	return f.info.DirectoryBlockInSeconds
+// SetTimeout sets a new timeout duration.
+// If the monitor is unable to connect to the factomd node within the duration,
+// an error is sent to all error listeners.
+func (f *Monitor) SetTimeout(timeout time.Duration) {
+	f.timeout = timeout
 }
 
-// Returns the highest saved block
-func (f *FactomdMonitor) GetHighestSavedDBlock() int64 {
-	f.mutex.Lock()
-	defer f.mutex.Unlock()
-	return f.info.DirectoryBlockHeight
+// getMinute queries the factomd api using exponential backoff.
+func (f *Monitor) getMinute() (*factom.CurrentMinuteInfo, error) {
+	bo := backoff.NewExponentialBackOff()
+	bo.InitialInterval = time.Millisecond * 100
+	bo.MaxElapsedTime = f.timeout
+	bo.MaxInterval = time.Second * 10
+	bo.Reset()
+
+	var info *factom.CurrentMinuteInfo
+	var err error
+	retry := func() error {
+		info, err = factom.GetCurrentMinute()
+		return err
+	}
+
+	fail := backoff.Retry(retry, bo)
+	if fail != nil {
+		return nil, fail
+	}
+	return info, err
```

Comment:
Why do you return that `err` here? If we reached that point it means that we successfully managed to the the current minute. Yet the `err` variable may be not nil if there was at least one failed retry, isn't it?

Reply:
yeah, it's always nil. updated!